tests/nodeos_producer_watermark_test.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python3
+
+from testUtils import Utils
+import testUtils
+from Cluster import Cluster
+from WalletMgr import WalletMgr
+from Node import Node
+from TestHelper import TestHelper
+
+import time
+import decimal
+import math
+import re
+
+###############################################################
+# nodeos_voting_test
+# --dump-error-details <Upon error print etc/eosio/node_*/config.ini and var/lib/node_*/stderr.log to stdout>
+# --keep-logs <Don't delete var/lib/node_* folders upon test completion>
+###############################################################
+class ProducerToNode:
+    map={}
+
+    @staticmethod
+    def populate(node, num):
```

Comment:
Why is map a static variable modified by a static function? Avoid globals. Make it an instance variable if necessary, or just define the function to mutate the map that is passed into the function as an argument.

Reply:
In fact, is this even necessary? It seems it is used only for extra information given in error messages and I'm not sure how much value it actually adds. 

That data is also not exactly accurate for shared producers because they belong to more than one node. 

I think it is totally reasonable to just print out the producers assigned to each node, then if someone is investigating the logs to debug an error, they can easily figure out which nodes they belonged to by looking it up in the assignments printed in the logs.